### ISSUE TITLE
P_Mask_Mo

### ISSUE THREAD

**Issue by 48-1-del** [2231122/SAM-COD#11]
Hello author, I'm curious about this step in your README: Filter and Matcher
Train P_Mask
Train the encoder and decoder supervised by Point_cu, scribble, or Max(SAM(Box_prompt)), where Max(SAM(Box_prompt)) is part of the Segment Anything model.

Put P_mask and segment anything's mask in './xxx'

python Filter_and_Matcher.py

How exactly is this P_Mask generated? I've been thinking about it for a long time and still can't figure it out. Filter_and_Matcher.py requires P_Mask as input, but the paper states that P_Mask is obtained by training the encoder and decoder. Could you please explain my confusion in more detail? Thank you!!!

**Comment by 2231122** [comment]
In fact, both Filter and Matcher are designed to improve the quality of the finally obtained mask. In other words, SAM can still generate masks without Filter and Matcher. The P_Mask used in the Matcher stage is acquired through training and subsequent inference based on the masks obtained in the Filter stage.

**Comment by 48-1-del** [comment]
I don't understand your reply: "The P_Mask used in the matcher stage is obtained through training and subsequent inference based on the mask obtained in the Filter stage." The Filter stage here corresponds to Filter_and_Matcher.py in the code. Here, P_Mask is required as input. Are you explaining that Filter_and_Matcher.py generates P_Mask?

**Comment by 2231122** [comment]
To put it simply: Use the masks processed by Filter to train E&D, then E&D performs inference to generate P_Mask, and finally the Matcher step is carried out. In the code, you need to first comment out the Matcher code and only execute the Filter step.

**Comment by 48-1-del** [comment]
You replied: Using the mask processed by the filter to train E&D, what is the code for "training E&D"? Is it train.py?

**Comment by 2231122** [comment]
yes

**Comment by 48-1-del** [comment]
I think there's a problem here. Firstly, the filter in Filter_and_Matcher.py is used to generate the filtered masks. Note that these are the inputs Masks_S_0/1/2, which can be generated by predictor_example_box.py. Note that the training set is used here. However, if you follow your statement: "Simply put: use the filtered masks to train E&D, then perform inference on E&D to generate P_Mask, and finally perform the matching step," then train E&D using train.py, then run the test file (test.py using the test set) to generate P_MASK. Finally, matching is performed, but P_MASK is generated from the test set, while Masks_S_0/1/2 are generated from the training set. Doesn't this mean a mismatch?

**Comment by 2231122** [comment]
Does using test.py for inference mean it must infer on the test set? Is there a possibility that this step itself infers on the training set to generate P_Mask (Pseudo-labeling)?

**Comment by 48-1-del** [comment]
From what I've learned, the test code shouldn't perform inference on the training set, otherwise how can we see the model's performance?

**Comment by 2231122** [comment]
P_Mask is generated by the trained E&D through inference on the training set.

**Comment by niu-yq** [comment]
> P_Mask is generated by the trained E&D through inference on the training set.

您好我想问一下，在这一个步骤中，使用train.py训练编解码器的时候，是不是要把里面相关的弱监督部分的代码都注释掉，包含dataset，data_prefetcher，train_processes等文件也需要改。
